File: internal/trace/trace.go
package trace

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"regexp"
	"strings"
	"time"
	"unicode"

	"gopkg.in/yaml.v3"
)

// ErrInvalidTraceID is returned when a trace ID contains characters that
// could escape the cortex directory (path traversal). Only IDs matching
// the YYYYMMDD-slug format produced by NewID are accepted.
var ErrInvalidTraceID = errors.New("invalid trace ID")

// validTraceID matches the format produced by NewID: 8 digits, a hyphen,
// then 1–100 characters of lowercase alphanumerics and hyphens. Pre-existing
// traces produced under the old 65-char cap remain valid — the regex is
// strictly a superset of the old pattern.
var validTraceID = regexp.MustCompile(`^[0-9]{8}-[a-z0-9][a-z0-9-]{0,99}$`)

// MaxSlugLen is the maximum length of the slug portion of a trace ID
// (the part after YYYYMMDD-). NewID truncates longer slugs. AGENTS.md
// documents this so agents can aim for short titles rather than
// discovering silent truncation after the fact.
const MaxSlugLen = 100

// IsValidID reports whether id is a well-formed trace ID safe for use in
// filesystem paths. It rejects any ID containing path separators, dot
// segments, or characters outside the NewID slug alphabet.
func IsValidID(id string) bool {
	return validTraceID.MatchString(id)
}

type Type string

const (
	TypeFact        Type = "fact"
	TypeDecision    Type = "decision"
	TypePreference  Type = "preference"
	TypeContext     Type = "context"
	TypeSkill       Type = "skill"
	TypeIntent      Type = "intent"
	TypeObservation Type = "observation"
	TypeNote        Type = "note"
	TypeDivergence  Type = "divergence"
)

var ValidTypes = []Type{
	TypeFact, TypeDecision, TypePreference, TypeContext,
	TypeSkill, TypeIntent, TypeObservation, TypeNote,
	TypeDivergence,
}

func IsValidType(t string) bool {
	for _, v := range ValidTypes {
		if string(v) == t {
			return true
		}
	}
	return false
}

// Tier names the three memory tiers in the consolidation hierarchy. See
// docs/plans/consolidation-plan.md (in the Noema-design repo) for the full
// design. New traces default to TierShort; promotion is short -> mid -> long.
// Long is terminal from routine operation (DB trigger enforces immutability).
const (
	TierShort = "short"
	TierMid   = "mid"
	TierLong  = "long"
)

var validTiers = map[string]struct{}{
	TierShort: {},
	TierMid:   {},
	TierLong:  {},
}

func IsValidTier(s string) bool {
	_, ok := validTiers[s]
	return ok
}

// ErrInvalidFrontmatter is returned by Validate when a Trace's frontmatter
// fails a required-field or enum check. The specific failure is wrapped
// underneath so callers can log it or branch on it.
var ErrInvalidFrontmatter = errors.New("invalid trace frontmatter")

// Validate checks that a parsed Trace satisfies the required-field and
// enum constraints the DB can't enforce on its own. The traces table
// declares id/title/type/created/updated as NOT NULL, but SQLite treats
// empty strings as satisfying NOT NULL, so YAML-missing fields would
// slip through as junk rows. Validate closes that gap for ingest paths
// that parse files written by external tools (the filesystem watcher,
// primarily — Cortex.Add and CLI flows build Traces via trace.New,
// which always fills these fields with sensible values).
//
// All errors wrap ErrInvalidFrontmatter so callers can use errors.Is
// to detect validation failures as a category.
func Validate(t *Trace) error {
	if t == nil {
		return fmt.Errorf("%w: nil trace", ErrInvalidFrontmatter)
	}
	if !IsValidID(t.ID) {
		return fmt.Errorf("%w: id %q does not match YYYYMMDD-slug format", ErrInvalidFrontmatter, t.ID)
	}
	if t.Title == "" {
		return fmt.Errorf("%w: title is required", ErrInvalidFrontmatter)
	}
	if t.Type == "" {
		return fmt.Errorf("%w: type is required", ErrInvalidFrontmatter)
	}
	if !IsValidType(t.Type) {
		return fmt.Errorf("%w: type %q is not one of the recognized types (fact, decision, preference, context, skill, intent, observation, note, divergence)", ErrInvalidFrontmatter, t.Type)
	}
	if t.Created == "" {
		return fmt.Errorf("%w: created timestamp is required", ErrInvalidFrontmatter)
	}
	if t.Updated == "" {
		return fmt.Errorf("%w: updated timestamp is required", ErrInvalidFrontmatter)
	}
	return nil
}

type Frontmatter struct {
	ID           string   `yaml:"id"`
	Title        string   `yaml:"title"`
	Type         string   `yaml:"type"`
	Tier         string   `yaml:"tier,omitempty"` // short|mid|long; empty == short, omitted for cleaner files
	Author       string   `yaml:"author,omitempty"`
	Tags         []string `yaml:"tags,omitempty"`
	DerivedFrom  []string `yaml:"derived_from,omitempty"`
	Origin       string   `yaml:"origin,omitempty"`
	Created      string   `yaml:"created"`
	Updated      string   `yaml:"updated"`
	ContentHash  string   `yaml:"content_hash,omitempty"`
	SourceHash   string   `yaml:"source_hash,omitempty"`
	SourceLocked bool     `yaml:"source_locked,omitempty"`
}

type Trace struct {
	Frontmatter
	Body string
}

func New(title, traceType, author string, tags []string, body string) *Trace {
	now := time.Now().UTC().Format(time.RFC3339)
	id := NewID(title)
	return &Trace{
		Frontmatter: Frontmatter{
			ID:      id,
			Title:   title,
			Type:    traceType,
			Author:  author,
			Tags:    tags,
			Created: now,
			Updated: now,
		},
		Body: body,
	}
}

func Parse(data []byte) (*Trace, error) {
	s := string(data)
	if !strings.HasPrefix(s, "---\n") {
		return nil, fmt.Errorf("missing frontmatter delimiter")
	}
	rest := s[4:]
	end := strings.Index(rest, "\n---\n")
	if end < 0 {
		return nil, fmt.Errorf("unterminated frontmatter")
	}
	fmStr := rest[:end]
	body := strings.TrimPrefix(rest[end+5:], "\n")

	var fm Frontmatter
	if err := yaml.Unmarshal([]byte(fmStr), &fm); err != nil {
		return nil, fmt.Errorf("parsing frontmatter: %w", err)
	}
	return &Trace{Frontmatter: fm, Body: body}, nil
}

func ParseFile(path string) (*Trace, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return Parse(data)
}

func (t *Trace) Write(path string) error {
	t.Frontmatter.Updated = time.Now().UTC().Format(time.RFC3339)
	fmData, err := yaml.Marshal(t.Frontmatter)
	if err != nil {
		return fmt.Errorf("encoding frontmatter: %w", err)
	}
	content := fmt.Sprintf("---\n%s---\n\n%s", string(fmData), t.Body)
	return os.WriteFile(path, []byte(content), 0o640)
}

// NewID generates a trace ID from a title: YYYYMMDD-slugified-title.
//
// If the slugified title already begins with a date prefix (either
// `YYYYMMDD-` or `YYYY-MM-DD-`), the leading date is stripped before
// today's date is prepended. This is a defensive measure: agents
// commonly include dates in titles to mark when an event occurred
// (e.g., "20260402 agent-1 heartbeat check"), and without the strip
// the resulting ID would carry two date prefixes
// (`20260402-20260402-agent-1-heartbeat-check`). The strip does not
// validate that the digits form a real calendar date — its only job
// is to prevent visual prefix duplication in the final ID.
func NewID(title string) string {
	date := time.Now().UTC().Format("20060102")
	s := stripLeadingDatePrefix(slug(title))
	if len(s) > MaxSlugLen {
		s = strings.TrimRight(s[:MaxSlugLen], "-")
	}
	return date + "-" + s
}

// stripLeadingDatePrefix removes a leading date prefix from s if one is
// present. It recognises two forms:
//
//	YYYYMMDD-     (8 digits + hyphen)         e.g. "20260402-foo" -> "foo"
//	YYYY-MM-DD-   (4-2-2 digits with hyphens) e.g. "2026-04-02-foo" -> "foo"
//
// Anything else — including titles where digits appear later in the
// string ("divergence-20260402-foo", "2026-was-a-year") — is returned
// unchanged.
func stripLeadingDatePrefix(s string) string {
	// YYYYMMDD- form (9 chars).
	if len(s) >= 9 && s[8] == '-' && allDigits(s[:8]) {
		return s[9:]
	}
	// YYYY-MM-DD- form (11 chars).
	if len(s) >= 11 && s[4] == '-' && s[7] == '-' && s[10] == '-' &&
		allDigits(s[:4]) && allDigits(s[5:7]) && allDigits(s[8:10]) {
		return s[11:]
	}
	return s
}

func allDigits(s string) bool {
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return false
		}
	}
	return true
}

// ContentHash returns the SHA-256 hash of a trace body, prefixed with
// "sha256:" for algorithm-explicitness. Used for integrity verification
// and federation sync optimization.
func ContentHash(body string) string {
	h := sha256.Sum256([]byte(body))
	return "sha256:" + hex.EncodeToString(h[:])
}

func slug(s string) string {
	s = strings.ToLower(s)
	var b strings.Builder
	prevHyphen := true
	for _, r := range s {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			b.WriteRune(r)
			prevHyphen = false
		} else if !prevHyphen {
			b.WriteByte('-')
			prevHyphen = true
		}
	}
	return strings.TrimRight(b.String(), "-")
}

File: rust/src/trace.rs
use std::{fs, path::Path, sync::LazyLock};

use anyhow::{Context, Result, bail};
use chrono::{SecondsFormat, Utc};
use regex::Regex;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

pub const MAX_SLUG_LEN: usize = 100;
pub const VALID_TYPES: &[&str] = &[
    "fact",
    "decision",
    "preference",
    "context",
    "skill",
    "intent",
    "observation",
    "note",
    "divergence",
];
pub const VALID_TIERS: &[&str] = &["short", "mid", "long"];

static VALID_ID: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^[0-9]{8}-[a-z0-9][a-z0-9-]{0,99}$").unwrap());

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Frontmatter {
    pub id: String,
    pub title: String,
    #[serde(rename = "type")]
    pub trace_type: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub tier: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub author: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub tags: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub derived_from: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub origin: String,
    pub created: String,
    pub updated: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub content_hash: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub source_hash: String,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub source_locked: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Trace {
    pub frontmatter: Frontmatter,
    pub body: String,
}

impl Trace {
    pub fn new(
        title: impl Into<String>,
        trace_type: impl Into<String>,
        author: impl Into<String>,
        tags: Vec<String>,
        body: impl Into<String>,
    ) -> Self {
        let title = title.into();
        let now = now_rfc3339();
        Self {
            frontmatter: Frontmatter {
                id: new_id(&title),
                title,
                trace_type: trace_type.into(),
                tier: String::new(),
                author: author.into(),
                tags,
                derived_from: Vec::new(),
                origin: String::new(),
                created: now.clone(),
                updated: now,
                content_hash: String::new(),
                source_hash: String::new(),
                source_locked: false,
            },
            body: body.into(),
        }
    }

    pub fn parse(data: &[u8]) -> Result<Self> {
        let source = std::str::from_utf8(data).context("trace is not UTF-8")?;
        let rest = source
            .strip_prefix("---\n")
            .ok_or_else(|| anyhow::anyhow!("missing frontmatter delimiter"))?;
        let end = rest
            .find("\n---\n")
            .ok_or_else(|| anyhow::anyhow!("unterminated frontmatter"))?;
        let frontmatter = serde_yaml::from_str(&rest[..end]).context("parsing frontmatter")?;
        let body = rest[end + 5..]
            .strip_prefix('\n')
            .unwrap_or(&rest[end + 5..]);
        Ok(Self {
            frontmatter,
            body: body.to_owned(),
        })
    }

    pub fn parse_file(path: &Path) -> Result<Self> {
        Self::parse(&fs::read(path).with_context(|| format!("reading {}", path.display()))?)
    }

    pub fn validate(&self) -> Result<()> {
        let f = &self.frontmatter;
        if !is_valid_id(&f.id) {
            bail!(
                "invalid trace frontmatter: id {:?} does not match YYYYMMDD-slug format",
                f.id
            );
        }
        if f.title.is_empty() {
            bail!("invalid trace frontmatter: title is required");
        }
        if !VALID_TYPES.contains(&f.trace_type.as_str()) {
            bail!(
                "invalid trace frontmatter: unrecognized type {:?}",
                f.trace_type
            );
        }
        if f.created.is_empty() || f.updated.is_empty() {
            bail!("invalid trace frontmatter: created and updated timestamps are required");
        }
        if !f.tier.is_empty() && !VALID_TIERS.contains(&f.tier.as_str()) {
            bail!("invalid trace frontmatter: unrecognized tier {:?}", f.tier);
        }
        Ok(())
    }

    pub fn encoded(&self) -> Result<Vec<u8>> {
        let yaml = serde_yaml::to_string(&self.frontmatter).context("encoding frontmatter")?;
        let yaml = yaml.strip_prefix("---\n").unwrap_or(&yaml);
        Ok(format!("---\n{yaml}---\n\n{}", self.body).into_bytes())
    }

    pub fn write(&mut self, path: &Path) -> Result<()> {
        self.frontmatter.updated = now_rfc3339();
        self.write_preserving_updated(path)
    }

    pub fn write_preserving_updated(&self, path: &Path) -> Result<()> {
        fs::write(path, self.encoded()?).with_context(|| format!("writing {}", path.display()))?;
        set_private_permissions(path)?;
        Ok(())
    }

    pub fn effective_tier(&self) -> &str {
        if self.frontmatter.tier.is_empty() {
            "short"
        } else {
            &self.frontmatter.tier
        }
    }
}

pub fn now_rfc3339() -> String {
    Utc::now().to_rfc3339_opts(SecondsFormat::Secs, true)
}

pub fn content_hash(body: &str) -> String {
    format!("sha256:{:x}", Sha256::digest(body.as_bytes()))
}

pub fn is_valid_id(id: &str) -> bool {
    VALID_ID.is_match(id)
}

pub fn new_id(title: &str) -> String {
    let date = Utc::now().format("%Y%m%d");
    let mut value = strip_leading_date_prefix(&slug(title)).to_owned();
    if value.chars().count() > MAX_SLUG_LEN {
        value = value.chars().take(MAX_SLUG_LEN).collect::<String>();
        while value.ends_with('-') {
            value.pop();
        }
    }
    format!("{date}-{value}")
}

fn slug(input: &str) -> String {
    let mut out = String::new();
    let mut previous_hyphen = true;
    for ch in input.to_lowercase().chars() {
        if ch.is_alphanumeric() {
            out.push(ch);
            previous_hyphen = false;
        } else if !previous_hyphen {
            out.push('-');
            previous_hyphen = true;
        }
    }
    out.trim_end_matches('-').to_owned()
}

fn strip_leading_date_prefix(value: &str) -> &str {
    let bytes = value.as_bytes();
    if bytes.len() >= 9 && bytes[8] == b'-' && bytes[..8].iter().all(u8::is_ascii_digit) {
        return &value[9..];
    }
    if bytes.len() >= 11
        && bytes[4] == b'-'
        && bytes[7] == b'-'
        && bytes[10] == b'-'
        && bytes[..4].iter().all(u8::is_ascii_digit)
        && bytes[5..7].iter().all(u8::is_ascii_digit)
        && bytes[8..10].iter().all(u8::is_ascii_digit)
    {
        return &value[11..];
    }
    value
}

#[cfg(unix)]
fn set_private_permissions(path: &Path) -> Result<()> {
    use std::os::unix::fs::PermissionsExt;
    fs::set_permissions(path, fs::Permissions::from_mode(0o640))?;
    Ok(())
}

#[cfg(not(unix))]
fn set_private_permissions(_path: &Path) -> Result<()> {
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn id_matches_go_rules() {
        let id = new_id("2026-04-02 Why We Chose Go!");
        assert!(id.ends_with("-why-we-chose-go"));
        assert!(is_valid_id(&id));
    }

    #[test]
    fn parse_and_encode_round_trip() {
        let mut trace = Trace::new("Choice", "decision", "tester", vec!["rust".into()], "body");
        trace.frontmatter.content_hash = content_hash(&trace.body);
        let parsed = Trace::parse(&trace.encoded().unwrap()).unwrap();
        assert_eq!(parsed, trace);
    }
}
